lib/mentionsPlugin.js

```diff
@@ -30,13 +30,19 @@ function parseMentions (state, silent) {
     return false;
   }
 
-  const text = state.src.slice(state.pos);
-  const match = text.match(mentionRegex);
+  const textUntilAt = state.src.slice(0, state.pos);
+  const textAfterAt = state.src.slice(state.pos);
 
-  if (!match || partOfLinkOrEmail(state.src.substr(0, state.pos), text, state.md)) {
+  const match = textAfterAt.match(mentionRegex);
+  if (!match) {
     return false;
   }
+
   const mention = match[0];
+  if (partOfLinkOrEmail(mention, textUntilAt, textAfterAt, state.md)) {
```

Comment:
Thanks @benkelaar for the careful feedback!

I now need `match[0]` in the call to `partOfLinkOrEmail`. The alternative is `if (!match && partOfLinkOrEmail(match[0], ...)` but in that case it's not clear what `match[0]` is without the variable name.

Reply:
Ah yeah, so it's a consideration between duplication of `match[0]` and duplication of `return false`.

I personally find that limiting the amount of separate possible exit-points of a function makes it easier to reason about program flow, which I think is worth the duplication of `match[0]` (especially since `if (!match || queryUsing(match.parameter))` is a fairly common pattern for null-checking in javascript).

The variable name is of course present as parameter name in `partOfLinkOrEmail`, which is where it is relevant.  However in the end it's a flavour call.

You could possibly turn the call into a call with [faux-named-parameters](https://exploringjs.com/impatient-js/ch_callables.html#named-parameters), but I don't think that'll increase the readability.  Might be a consideration though.

Reply:
Hmm, now that I think about it though, that would turn it in to:

```
function partOfLinkOrEmail ({ mention, textUntilAt, textAfterAt, md }) {
// ...

// and then later
if (!match || partOfLinkOrEmail({mention: match[0], textUntilAt, textAfterAt, md: state.md}))
```

Which might just be exactly what you need. Haven't tried it yet though, perhaps you don't even need to provide the name for `state.md` either, but not sure.

Reply:
The linter was not fond of that named parameter approach without line breaks before/after the curly braces, which looks hideous. I'll switch to the `match[0]` approach since that is more idiomatic. FWIW, I have been mostly coding in Go lately, where early exit on edge cases (leaving only the happy path) is considered idiomatic and more readable.